src/main/cpp/src/slice.cu

```diff
@@ -0,0 +1,358 @@
+                                      if (start.is_null(index)) return true;
+                                      return start.element<int32_t>(index) != 0;
+                                    });
+  CUDF_EXPECTS(start_valid, "Invalid start value: start must not be 0");
+}
+
+void assert_length_is_not_negative(column_device_view const& length, rmm::cuda_stream_view stream)
+{
+  bool length_valid = thrust::all_of(rmm::exec_policy(stream),
+                                     thrust::make_counting_iterator(0),
+                                     thrust::make_counting_iterator(length.size()),
+                                     [length] __device__(size_type index) {
+                                       if (length.is_null(index)) return true;
+                                       return length.element<int32_t>(index) >= 0;
+                                     });
+  CUDF_EXPECTS(length_valid, "Invalid length value: length must be >= 0");
+}
+
+struct int_iterator_from_scalar {
+  int32_t value;
+  int_iterator_from_scalar(int32_t v) : value(v) {}
+  __device__ bool is_null(cudf::thread_index_type const index) const { return false; }
+  __device__ int32_t operator()(cudf::thread_index_type const index) const
+  {
+    // ignore index, always return the same value
+    return value;
+  }
+};
+
+struct int_iterator_from_column {
+  column_device_view cdv;
+  int_iterator_from_column(column_device_view cdv) : cdv(cdv) {}
+  __device__ bool is_null(cudf::thread_index_type const index) const { return cdv.is_null(index); }
+  __device__ int32_t operator()(cudf::thread_index_type const index) const
+  {
+    return cdv.element<int32_t>(index);
+  }
+};
+
+template <typename SRART_ITERATOR, typename LENGTH_ITERATOR>
+CUDF_KERNEL void compute_starts_and_sizes_kernel(size_type const* offsets_of_input_lists,
+                                                 size_type const num_rows,
+                                                 SRART_ITERATOR const start_iterator,
+                                                 LENGTH_ITERATOR const length_iterator,
+                                                 size_type* d_starts,
+                                                 size_type* d_sizes)
+{
+  auto const tid = cudf::detail::grid_1d::global_thread_id();
+  if (tid >= num_rows) { return; }
+
+  if (start_iterator.is_null(tid) || length_iterator.is_null(tid)) {
+    d_sizes[tid] = 0;
+    return;
+  }
+
+  auto const length_of_list = offsets_of_input_lists[tid + 1] - offsets_of_input_lists[tid];
+  auto start                = start_iterator(tid);
+  auto const length         = length_iterator(tid);
+
+  start = start < 0 ? length_of_list + start : start - 1;
+  if (start < 0 || start >= length_of_list) {
+    d_sizes[tid] = 0;
+    return;
+  }
+  d_starts[tid] = start;
+  d_sizes[tid]  = cuda::std::min(length_of_list - start, length);
+}
+
+CUDF_KERNEL void compute_gather_map(size_type const num_rows_of_input,
```

Comment:
better to have comments to describe its functionality and the input and output for easier maintenance in the future.